### DISCUSSION TITLE
mercantile replacement utiles!

### DISCUSSION THREAD

**Discussion by jessekrubin** [developmentseed/morecantile#126] [General]
Hello Morecantile team!

I have used a bunch of `mercantile` in my time, and as a side-project, I have ported (and published) a mostly drop-in replacement for `mercantile` called `utiles`. It is written with Python + Rust; I used the project as a way of learning some rust. You can check it out at https://github.com/jessekrubin/utiles! It seems to be faster than `mercantile`, although I believe there might be more optimizations that I am not aware of.

Differences with mercantile:
 - `utiles` raises `TypeErrors` and `ValueError`s instead of custom errors
 - `utiles` adds more object oriented functionality to the struct/tuple object types
 - `utiles` adds utils for getting/parsing pmtile-id (based on the hilbert curve)

I'm wondering if other Mercantile aficionados would be interested in checking out utiles and possibly contributing to it!

I would be happy to add any `morecantile` functionality to utiles as well.

Thank you,
Jesse

PS (edit): it adds type annotations too!

**Comment by vincentsarago** [comment]
🥳 that's great @jessekrubin 

FYI, @pka also created a port of Morcantile in Rust: https://github.com/pka/tile-grid

**Reply by jessekrubin** [reply]
@vincentsarago sounds like we (@pka and you) should work together!

**Reply by pka** [reply]
My port is not intended to use in Python. Its main purpose is being an idiomatic Rust library for OGC API Tiles implementations.

**Reply by jessekrubin** [reply]
@pka any which way I would be down to work w/ you!

**Reply by jessekrubin** [reply]
@pka if you look at the rust src for utiles you will see I split up the rust stuff into `utiles` and `pyutiles` and was thinking of separating and publishing the rust only version as a library-crate. Not sure what is involved but curious what you think!

PS: Also am curious what you think of the actual rust src and if you see any glaring things I missed or should change! Still learning rust!

**Reply by pka** [reply]
I just realized, that I've read *more*cantile instead of *mer*cantile! My primary goal was to support non-Mercator projections for tile servers. As a native Rust alternative to Proj, I've implemented the Mercator projection, but eventually want to use https://github.com/busstoptaktik/geodesy/

I still have some ideas for collaboration. Since you're obviously interested in performance, a few Rust benches (like e.g. https://github.com/georust/geozero/tree/main/geozero-bench/benches) to compare utiles with tile-grid would be great to have. And I also saw some `From` implementations, tile-grid would benefit from. If you could add a compatible license to utiles, we could also share code.

**Reply by jessekrubin** [reply]
@pka alrighty! I was able to extract the package to a crate (branch: https://github.com/jessekrubin/utiles/tree/crate) This is my first publich rust thing so lmk if you have any suggestions! I based the cargo-toml metadata off your library!

https://crates.io/crates/utiles publishing the rust only part was alarmingly easy!

Happy to add benchmarks, but don't know how to do that in rust alone, so guidance appreciated! Also more than happy to accept a PR from you or anyone else! 

I will be around for messages, but I must return to my day-job for the rest o the day! (which if you were curious is geo-science-computing stuff)

**Comment by vincentsarago** [comment]
FYI, When we'll switch to pydantic V2, we might want to re-write morecantile in Rust (ping @kylebarron ;-) )

**Reply by jessekrubin** [reply]
that is going to be a funky/gnarly upgrade! I am not looking forward to doing the upgrade for my stuff (which (if you were curious) is all in a package called `jsonbourne`)
